en/material-system/effect-syntax.md

```diff
@@ -0,0 +1,259 @@
+vec4 vert () {
+  StandardVertInput In;
+  CCVertInput(In);
+  // ... now use `In.position`, etc.
+}
+```
+This will acquire position, normal and tangent data, all after vertex skinning.<br>
+Other vertex data (uv, color, etc.) can be declared directly.
+
+To support dynamic batching, use `CCGetWorldMatrix`:
+```glsl
+#include <cc-local-batch>
+
+vec4 vert () {
+  // ...
+  // unlit version (when normal is not needed)
+  mat4 matWorld;
+  CCGetWorldMatrix(matWorld);
+  // standard version
+  mat4 matWorld, matWorldIT;
+  CCGetWorldMatrixFull(matWorld, matWorldIT);
+  // ...
+}
+```
+
+### Fragment Ouput<sup id="a1">[1](#f1)</sup>
+To encapsulate render pipeline complexities, use `CCFragOutput`:
+For unlit shaders:
+```glsl
+#include <output>
+vec4 frag () {
+  vec4 o = vec4(0.0);
+  // ... do the computation
+  return CCFragOutput(o);
+}
+```
+So the code can work in both HDR and LDR pipeline.<br>
+If there's lighting involved, combine with `CCStandardShading` to form a surface shader structure:
+```glsl
+#include <shading-standard>
+#include <output-standard> // note the header file change
+void surf (out StandardSurface s) {
+  // fill in your data here
+}
+vec4 frag () {
+  StandardSurface s; surf(s);
+  vec4 color = CCStandardShading(s);
+  return CCFragOutput(color);
+}
+```
+Under the framework writing your own surface shader or even shading algorithm becomes staightforward;<br>
+Note that the `CCFragOutput` function should not be overriden, unless using custom render pipelines.
+
+### WebGL 1 fallback Support
+The effect compiler provides fallback conversion from GLSL 300 ES to GLSL 100 automatically, for WebGL 1.0 only support GLSL 100 syntax. this should be transparent to developers for the most time.<br>
+But currently the automatic conversion only supports some basic usage, and if some post-100 features or extensions were used, (texelFetch, textureGrad, etc.)<br>
+you have to do your own porting using the language built-in \_\_VERSION__ macro:
+```glsl
+vec4 fragTextureLod (samplerCube tex, vec3 coord, float lod) {
+  #if __VERSION__ < 130
+    #ifdef GL_EXT_shader_texture_lod
+      return textureCubeLodEXT(tex, coord, lod);
+    #else
+      return textureCube(tex, coord, lod); // fallback to bias
+    #endif
+  #else
+    return textureLod(tex, coord, lod);
+  #endif
+}
+```
+The effect compiler will finally split these compile-time constant branches into different output versions.
+
+### About UBO Memory Layout
+First the conclusion, the final rules are, every non-sampler uniform should be specified in UBO blocks, and for every UBO block:
+* there should be no vec3 typed members；
+* for array typed members, size of each element should be no less than a vec4;
+* any member ordering that introduces a padding will be rejected;
+
+These rules will be checked rigorously at effect compile-time, and throws detailed,implicit padding related compile error.
+
+This might sounds overly-strict at first, but for good reasons:<br>
```

Comment:
This might sounds overly-strict at first, but for good reasons:<br>
-->
This might sound overly-strict at first, but it's for a few good reasons:<br>